docs/proposals/010-new-diff-type-system.md

```diff
@@ -0,0 +1,283 @@
+{
+	// Force arithmetic types to be a differential pair of the same two types.
+	// Make it simple to define derivatives of arithmetic operations.
+	//
+	associatedtype Differential : This;
+}
+
+extension IFloat : ISelfDifferentiableValueType
+{ }
+
+extension float
+{
+	// trivial auto-synthesis (maybe we even prevent the user from overriding this)
+	float use() { return this; } 
+	
+	// trivial auto-synthesis (maybe we even prevent the user from overriding this).
+	[ForwardDerivativeOf(use)]
+	[deriv_method] void use_fwd() { return this; } 
+	
+	// auto-synthesized if necessary by invoking the use_bwd for all fields.
+	// we need to provide implementation for 'leaf' types.
+	[BackwardDerivativeOf(use)]
+	[deriv_method] [mutating] void use_bwd(float d) { this.d += d; } 
+}
+
+// The new system lets us define differentiable pointers easily.
+// IDifferentiablePtrType'd values are simply treated as references, so they can be freely
+// duplicated without requiring a `use()` for correctness.
+//
+struct DPtr<T : IDifferentiableValueType> : IDifferentiablePtrType
+{
+	typealias Differential = DPtr<T.Differential>;
+	
+	Buffer<T> buffer;
+	uint64 offset;
+	
+	[BackwardDerivative(get_bwd)]
+	[BackwardDerivative(get_fwd)]
+	T get() { return this.buffer[offset]; }
+	
+	[deriv_method] DifferentialPair<T> get_fwd()
+	{ 
+		return diffPair(this.p.buffer[offset], this.d().buffer[offset]);
+	}
+	
+	[deriv_method] void get_bwd(Differential d)
+	{
+		return this.d.InterlockedAdd(offset, d);
+	}
+	
+	DPtr<T> operator+(uint o) { return DPtr<T>{buffer, offset + o}; }
+}
+
+// Or we can define a fancier differentiable pointer that does a hashgrid
+struct DHashGridPtr<T : IDifferentiableValueType, let N: int> : IDifferentiablePtrType
+{ 
+	typealias Differential = DPtr<T.Differential>;
+	
+	Buffer<T> buffer;
+	uint64 offset;
+	
+	[BackwardDerivative(get_bwd)]
+	[BackwardDerivative(get_fwd)]
+	T get() { return this.buffer[offset]; }
+	
+	[deriv_method] DifferentialPair<T> get_fwd()
+	{ 
+		return diffPair(this.p().buffer[offset], this.d().buffer[offset]);
+	}
+	
+	[deriv_method] void get_bwd(Differential d)
+	{
+		return this.d().InterlockedAdd(offset * N + hash(get_thread_id()), d);
+	}
+}
+```
+
+### 3. Every time we 'reuse' an object that conforms to `IDifferentiableValueType`, we split it with `use()` , and we use `__init__()` where necessary to initialize an accumulator.
+Example:
+```csharp
+float f(float a)
+{
+	add(a, a);
+}
+float add(float a, float b)
+{
+	return a + b;
+}
+
+// Synthesized derivatives
+void add_bwd(inout DiffPair<float> dpa, inout DiffPair<float> dpb, float d_out)
+{
+	 dpa = diffPair(dpa.p, d_out);
+	 dpb = diffPair(dpb.p, d_out);
+}
+
+// Preprocessed-f (before derivative generation)
+float f_with_use_expansion(float a)
+{
+	DiffPair<float> a_extra = a.use();
+	return add(a, a_extra);
+}
+
+// After fwd-mode:
+DiffPair<float> f_fwd(DiffPair<float> dpa)
+{
+	DiffPair<float> dpa_extra = dpa.use_fwd();
```

Comment:
what does `DiffPair<float>.use_fwd` do?

Reply:
It is effectively a noop 'passthrough' in this case. It only exists because when converting to reverse-mode, we will replace `use_fwd` into `use_bwd` which is the method that has non-trivial logic. Both `use` and `use_fwd` are always intended to be passthroughs. 
(We could do away with `use` entirely and simply require the implementation of the backward part, which is just `dadd` but not static)